### ISSUE TITLE
`todo!` macro does not satisfy type-checker when `impl Trait` is in the return position

### ISSUE THREAD

**Issue by duncpro** [rust-lang/rust#132674]
The `todo!` macro does not satisfy the type-checker when `impl Trait` is in the return position. 

This behavior contradicts the official documentation.

> [todo!] can be useful if you are prototyping and just want a placeholder to let your code pass type analysis.
>
> \- [doc for std::todo macro](https://doc.rust-lang.org/std/macro.todo.html)

## Example

```rust
// src/main.rs
struct T;

fn make_iter() -> impl Iterator<Item = T> {
    todo!()
}

fn main() {
    make_iter();
}
```

```shell
cargo run
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9eca88daedb0cd01da9cd5af13b3b9dd)

### Expected Behavior
The code to compile normally and panic during execution.

### Actual Behavior
The code fails to compile because it does not pass typecheck.
![Image](https://github.com/user-attachments/assets/5c4e7004-b120-4279-b8ee-5cabd8e74add)

## `rustc --version --verbose`:
```
rustc 1.82.0 (f6e511eec 2024-10-15)
binary: rustc
commit-hash: f6e511eec7342f59a25f7c0534f1dbea00d01b14
commit-date: 2024-10-15
host: aarch64-apple-darwin
release: 1.82.0
LLVM version: 19.1.1
```

**Comment by theemathas** [comment]
Duplicate of #113875

**Comment by duncpro** [comment]
Perhaps a note can be added to the docs for `todo!` so this doesn't get reported again?
